modules/tensor_mechanics/src/interfacekernels/CZMInterfaceKernelBase.C

```diff
@@ -0,0 +1,108 @@
+  params.addRequiredParam<unsigned int>("component",
+                                        "the component of the "
+                                        "displacement vector this kernel is working on:"
+                                        " component == 0, ==> X"
+                                        " component == 1, ==> Y"
+                                        " component == 2, ==> Z");
+  params.set<bool>("_use_undisplaced_reference_points") = true;
+  params.suppressParameter<bool>("use_displaced_mesh");
+  params.addRequiredCoupledVar("displacements", "the string containing displacement variables");
+  params.addParam<std::string>("base_name", "Material property base name");
+  params.set<std::string>("traction_global_name") = "traction_global";
+
+  return params;
+}
+
+CZMInterfaceKernelBase::CZMInterfaceKernelBase(const InputParameters & parameters)
+  : JvarMapKernelInterface<InterfaceKernel>(parameters),
+    _base_name(isParamValid("base_name") && !getParam<std::string>("base_name").empty()
+                   ? getParam<std::string>("base_name") + "_"
+                   : ""),
+    _component(getParam<unsigned int>("component")),
+    _ndisp(coupledComponents("displacements")),
+    _disp_var(_ndisp),
+    _disp_neighbor_var(_ndisp),
+    _vars(_ndisp),
+    _traction_global(getMaterialPropertyByName<RealVectorValue>(
+        _base_name + getParam<std::string>("traction_global_name"))),
+    _dtraction_djump_global(
+        getMaterialPropertyByName<RankTwoTensor>(_base_name + "dtraction_djump_global"))
+{
+  // Enforce consistency
+  if (_ndisp != _mesh.dimension())
+    paramError("displacements", "Number of displacements must match problem dimension.");
+
+  if (_ndisp > 3 || _ndisp < 1)
+    mooseError("the CZM material requires 1, 2 or 3 displacement variables");
+}
+
+void
+CZMInterfaceKernelBase::initialSetup()
+{
+  for (unsigned int i = 0; i < _ndisp; ++i)
+  {
+    _disp_var[i] = coupled("displacements", i);
+    _disp_neighbor_var[i] = coupled("displacements", i);
+    _vars[i] = getVar("displacements", i);
+  }
+}
+
+Real
+CZMInterfaceKernelBase::computeQpResidual(Moose::DGResidualType type)
+{
+
+  Real r = _traction_global[_qp](_component);
+
+  switch (type)
+  {
+    // [test_secondary-test_primary]*T where T represents the traction.
+    case Moose::Element:
+      r *= -_test[_i][_qp];
+      break;
+    case Moose::Neighbor:
+      r *= _test_neighbor[_i][_qp];
+      break;
+  }
+  return r;
+}
+
+Real
+CZMInterfaceKernelBase::computeQpJacobian(Moose::DGJacobianType type)
+{
+  // retrieve the diagonal Jacobian coefficient dependning on the displacement
+  // component (_component) this kernel is working on
+  return computeDResidualDDisplacement(_component, type);
+}
+
+Real
+CZMInterfaceKernelBase::computeQpOffDiagJacobian(Moose::DGJacobianType type, unsigned int jvar)
+{
+  Real jac = 0.0;
+  // bail out if jvar is not coupled
+  if (getJvarMap()[jvar] < 0)
+    return jac;
```

Comment:
I am not familiar with this function. Do we really need this?

Reply:
this function catches if a variable is coupled with the kernel or not. If it is not it returns a negative value. This is how the `StressDiveregenceKernel` works, as per you previous comment